File: Leetcode-Python/InvertBinaryTree.py
"""
Invert a binary tree.

     4
   /   \
  2     7
 / \   / \
1   3 6   9
to
     4
   /   \
  7     2
 / \   / \
9   6 3   1
"""


class Solution:
    # @param {TreeNode} root
    # @return {TreeNode}
    def invertTree(self, root):
        if root:
            root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)
        return root

    def invertTreeBFS(self, root):
        if not root:
            return root
        q = list()
        q.append(root)
        while q:
            size = len(q)
            for _ in xrange(size):
                node = q.pop()
                node.left, node.right = node.right, node.left
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        return root

File: Java-By-Tag/src/main/InvertBinaryTree.java
package src.main;

/**
 * Created by yhf on 10/6/15.
 */

import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

/**
 * Invert a binary tree.
 */

public class InvertBinaryTree {
    public TreeNode invertTree(TreeNode root) {
        // DFS
        if (root != null) {
            TreeNode temp = root.left;
            root.left = root.right;
            root.right = temp;
            invertTree(root.left);
            invertTree(root.right);
        }
        return root;
    }

    public TreeNode invertTreeBFS(TreeNode root) {
        // BFS
        if (root == null) {
            return root;
        }
        List<TreeNode> q = new ArrayList<>();
        q.add(root);
        while (!q.isEmpty()) {
            int size = q.size();
            while (size > 0) {
                size--;
                TreeNode node = q.remove(0);
                if (node != null) {
                    TreeNode temp = node.left;
                    node.left = node.right;
                    node.right = temp;
                    q.add(node.right);
                    q.add(node.left);
                }
            }
        }
        return root;
    }
}
